File: backend/src/api/shared/submit.rs
use serde::{Deserialize, Serialize};
use strum::{EnumString, IntoStaticStr};

#[derive(
    Debug, Clone, Copy, PartialEq, Eq, Hash, IntoStaticStr, EnumString,
)]
pub enum Status {
    Rejected,
    Passed,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum PreSubmitStatus {
    Rejected {
        reason: PreSubmitRejectReason,
    },
    Passed {
        lead: bool,
        choir: bool,
        harmony: bool,
    },
}

#[derive(
    Debug,
    Clone,
    Copy,
    PartialEq,
    Eq,
    Hash,
    IntoStaticStr,
    EnumString,
    Serialize,
    Deserialize,
)]
pub enum PreSubmitRejectReason {
    DeviceOrEnvironment,
    RequirementNotMet,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum SubmitStatus {
    Rejected {
        reason: PreSubmitRejectReason,
        detail: Option<Box<str>>,
    },
    Passed,
}

#[derive(
    Debug,
    Clone,
    Copy,
    PartialEq,
    Eq,
    Hash,
    IntoStaticStr,
    EnumString,
    Serialize,
    Deserialize,
)]
pub enum SubmitRejectReason {
    DeviceOrEnvironment,
    RequirementNotMet,
    Other,
}

File: manage/src/shared/submit.ts
export type GroupInfo = {
  lead: boolean;
  choir: boolean;
  harmony: boolean;
  choir_harmony: boolean;
};

export const groupInfoTxt: {
  [K in keyof GroupInfo]: string;
} = {
  lead: "领唱",
  choir: "合唱",
  harmony: "和声",
  choir_harmony: "合唱和声",
};

export type Detail<S> = {
  mid: number;
  mname: string;
  status: S;
};

export type PreSubmitStatus =
  | { "Rejected": { reason: PreSubmitRejectReason } }
  | { "Passed": GroupInfo };

export type PreSubmitRejectReason =
  | "DeviceOrEnvironment"
  | "RequirementNotMet"
  | "InvalidName";

export const preSubmitRejectReasonTxt: {
  [T in PreSubmitRejectReason]: string;
} = {
  DeviceOrEnvironment: "设备或环境",
  RequirementNotMet: "未达到标准",
  InvalidName: "用户名问题",
};

export type SubmitStatus =
  | { "Rejected": { reason: SubmitRejectReason; detail: null | string } }
  | "Passed";

export type SubmitRejectReason =
  | "DeviceOrEnvironment"
  | "RequirementNotMet"
  | "Other";

export const submitRejectReasonTxt: {
  [T in SubmitRejectReason]: string;
} = {
  DeviceOrEnvironment: "设备或环境",
  RequirementNotMet: "未达到标准",
  Other: "其他",
};
